File: algorithms/algorithms-cplusplus/leetcode/Question_0036_Valid_Sudoku.cc
#include <vector>

using namespace std;

class Solution {
 public:
  bool isValidSudoku(vector<vector<char>>& board) {
    size_t m = board.size(), n = board[0].size();
    vector<vector<bool>> row(m, vector<bool> (n,false));
    vector<vector<bool>> col(m, vector<bool> (n,false));
    vector<vector<bool>> area(m, vector<bool> (n,false));
    for (size_t i = 0; i < m; i++) {
      for (size_t j = 0; j < n; j++) {
        if(board[i][j] == '.') {
          continue;
        }
        size_t k = board[i][j] - '1';
        size_t p = i/3*3 + j/3;
        if (row[i][k] || col[j][k] || area[p][k]) {
          return false;
        }
        row[i][k] = col[j][k] = area[p][k] = true;
      }
    }
    return true;
  }
};

File: algorithms/algorithms-python/leetcode/Question_0036_Valid_Sudoku.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        row = [[False for _ in range(9)] for _ in range(9)]
        col = [[False for _ in range(9)] for _ in range(9)]
        area = [[False for _ in range(9)] for _ in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    continue
                
                k = int(board[i][j]) - 1
                p = i//3*3 + j//3;
                if row[i][k] or col[j][k] or area[p][k]:
                    return False

                row[i][k], col[j][k], area[p][k] = True, True, True
        return True
